packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+import {NativePath, PortablePath}          from '@yarnpkg/fslib';
+import {parse, init}                       from 'cjs-module-lexer';
+import fs                                  from 'fs';
+import moduleExports                       from 'module';
+import path                                from 'path';
+import {fileURLToPath, pathToFileURL, URL} from 'url';
+
+import * as nodeUtils                      from '../loader/nodeUtils';
+import {PnpApi}                            from '../types';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...moduleExports.builtinModules]);
+
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:node:)?(?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+async function exists(path: string) {
+  try {
+    await fs.promises.access(path, fs.constants.R_OK);
+    return true;
+  } catch { }
+  return false;
+}
```

Comment:
I'd rather promisify `fs.exists` - if I recall correctly, `exists` using a try/catch was adding a slight but notable overhead.

Reply:
`fs.exists` is deprecated.

Reply:
It's also just a wrapper for `fs.access` as well so the performance should be the same
https://github.com/nodejs/node/blob/ce666d769da67e7526e3084b661c011f534ea67a/lib/fs.js#L250-L274